x/gmp/keeper/keeper.go

```diff
@@ -58,41 +59,65 @@ func (k Keeper) RelayPrice(
 	ctx := sdk.UnwrapSDKContext(goCtx)
 	params := k.GetParams(ctx)
 
-	// encode oracle data
-	rates := []types.PriceFeedData{}
+	prices := []types.PriceData{}
 	for _, denom := range msg.Denoms {
+		// get exchange rate
 		rate, err := k.oracleKeeper.GetExchangeRate(ctx, denom)
 		if err != nil {
 			return &types.MsgRelayPriceResponse{}, err
 		}
 
-		priceFeed, err := types.NewPriceFeedData(
+		// get any available median and standard deviation data
+		medians := k.oracleKeeper.HistoricMedians(
+			ctx,
+			denom,
+			k.oracleKeeper.MaximumMedianStamps(ctx),
+		)
+		deviations := k.oracleKeeper.HistoricDeviations(
+			ctx,
+			denom,
+			k.oracleKeeper.MaximumMedianStamps(ctx),
+		)
+		// convert them to a medianData slice
+		medianData, err := types.NewMediansSlice(medians, deviations)
+		if err != nil {
+			return &types.MsgRelayPriceResponse{}, err
+		}
+
+		priceFeed, err := types.NewPriceData(
 			denom,
 			rate,
-			// TODO: replace with actual resolve time & id
-			// Ref: https://github.com/ojo-network/ojo/issues/309
-			big.NewInt(1),
-			big.NewInt(1),
+			big.NewInt(msg.Timestamp),
+			medianData,
 		)
 		if err != nil {
 			k.Logger(ctx).With(err).Error("unable to relay price to gmp")
 			continue
 		}
 
-		rates = append(rates, priceFeed)
+		prices = append(prices, priceFeed)
 	}
 
-	// TODO: fill with actual disableResolve option
-	// Ref: https://github.com/ojo-network/ojo/issues/309
-	payload, err := types.EncodeABI("postPrices", rates, false)
+	// convert commandSelector to [4]byte
+	var commandSelector [4]byte
+	copy(commandSelector[:], msg.CommandSelector)
+
+	encoder := types.NewGMPEncoder(
+		prices,
+		msg.Denoms,
+		common.HexToAddress(msg.ContractAddress),
+		commandSelector,
+		msg.CommandParams,
+	)
+	payload, err := encoder.GMPEncode()
 	if err != nil {
 		return nil, err
 	}
 
 	// package GMP
 	message := types.GmpMessage{
 		DestinationChain:   msg.DestinationChain,
-		DestinationAddress: msg.DestinationAddress,
+		DestinationAddress: msg.ContractAddress,
```

Comment:
do these two need to have separate contractAddresses?

My thought is that the encoder uses the "client" contract address, while the GmpMessage struct uses the "ojo" contract address?

I'm just not sure how the evm interface works out here - @rbajollari ?

Reply:
Yeah this DestinationAddress should be the Ojo Contract address while the payload contains the client contract address (so they are different)

Reply:
Resolved here: 15f7e39e526d0df817ef792abb847a056b01ec89